File: python/horus/experiments/metrics.py
"""Metrics writers used by HORUS benchmark sources and analysis tools."""

from __future__ import annotations

import csv
from dataclasses import asdict, is_dataclass
import json
import os
from pathlib import Path
import time
from typing import Any, Dict, Iterable, Mapping, Optional, Sequence


COMMON_FIELDS = ("timestamp_ns", "run_id", "experiment", "condition")
CLOCK_WALL_ANCHOR_ENV = "HORUS_EXPERIMENT_CLOCK_WALL_ANCHOR_NS"
CLOCK_MONOTONIC_ANCHOR_ENV = "HORUS_EXPERIMENT_CLOCK_MONOTONIC_ANCHOR_NS"


def _env_int(name: str, fallback: int) -> int:
    try:
        return int(os.getenv(name, ""))
    except (TypeError, ValueError):
        return fallback


_CLOCK_WALL_ANCHOR_NS = _env_int(CLOCK_WALL_ANCHOR_ENV, time.time_ns())
_CLOCK_MONOTONIC_ANCHOR_NS = _env_int(CLOCK_MONOTONIC_ANCHOR_ENV, time.monotonic_ns())


def clock_anchor_env() -> Dict[str, str]:
    return {
        CLOCK_WALL_ANCHOR_ENV: str(_CLOCK_WALL_ANCHOR_NS),
        CLOCK_MONOTONIC_ANCHOR_ENV: str(_CLOCK_MONOTONIC_ANCHOR_NS),
    }


def now_ns() -> int:
    """Return a Unix-epoch-like timestamp derived from a monotonic clock.

    WSL and VM environments can step wall-clock time while an experiment is
    running. The benchmark needs timestamps that share an epoch with Quest/Unity
    wall time but remain monotonic within the host process, so anchor once to
    wall time and advance using ``time.monotonic_ns()``.
    """
    return _CLOCK_WALL_ANCHOR_NS + (time.monotonic_ns() - _CLOCK_MONOTONIC_ANCHOR_NS)


def monotonic_ns() -> int:
    return time.monotonic_ns()


def default_metrics_path(file_name: str) -> Optional[Path]:
    root = os.getenv("HORUS_EXPERIMENT_RESULTS_DIR")
    if not root:
        return None
    return Path(root) / file_name


def source_metric_stream_seen(path: Path, stream: str) -> bool:
    path = Path(path)
    if not path.exists():
        return False
    try:
        with path.open("r", encoding="utf-8", newline="") as handle:
            for row in csv.DictReader(handle):
                if str(row.get("stream") or "") == stream:
                    return True
    except Exception:
        return False
    return False


def normalize_row(row: Any) -> Dict[str, Any]:
    if is_dataclass(row):
        row = asdict(row)
    if isinstance(row, Mapping):
        return dict(row)
    raise TypeError(f"Metric rows must be mappings or dataclasses, got {type(row)!r}")


class CsvMetricWriter:
    """Append metrics to a CSV with stable common fields.

    The writer intentionally accepts dictionaries so benchmark scripts can add
    experiment-specific columns without changing the core package.
    """

    def __init__(
        self,
        path: Path,
        *,
        run_id: str,
        experiment: str,
        condition: str,
        fieldnames: Sequence[str],
    ) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.run_id = run_id
        self.experiment = experiment
        self.condition = condition
        self.fieldnames = self._merge_fields(fieldnames)
        self._file = self.path.open("w", newline="", encoding="utf-8")
        self._writer = csv.DictWriter(self._file, fieldnames=self.fieldnames, extrasaction="ignore")
        self._writer.writeheader()

    def _merge_fields(self, fieldnames: Sequence[str]) -> Sequence[str]:
        fields = list(COMMON_FIELDS)
        for field in fieldnames:
            if field not in fields:
                fields.append(field)
        return fields

    def write(self, row: Any, *, timestamp_ns: Optional[int] = None) -> None:
        data = normalize_row(row)
        data.setdefault("timestamp_ns", timestamp_ns if timestamp_ns is not None else now_ns())
        data.setdefault("run_id", self.run_id)
        data.setdefault("experiment", self.experiment)
        data.setdefault("condition", self.condition)
        self._writer.writerow(data)

    def writerows(self, rows: Iterable[Any]) -> None:
        for row in rows:
            self.write(row)

    def flush(self) -> None:
        self._file.flush()

    def close(self) -> None:
        self._file.flush()
        self._file.close()

    def __enter__(self) -> "CsvMetricWriter":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()


class NdjsonEventWriter:
    """Write event records as newline-delimited JSON."""

    def __init__(
        self,
        path: Path,
        *,
        run_id: str,
        experiment: str,
        condition: str,
        source: str,
    ) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.run_id = run_id
        self.experiment = experiment
        self.condition = condition
        self.source = source
        self._file = self.path.open("w", encoding="utf-8")

    def write(self, event: Mapping[str, Any], *, timestamp_ns: Optional[int] = None) -> None:
        payload: Dict[str, Any] = dict(event)
        payload.setdefault("timestamp_ns", timestamp_ns if timestamp_ns is not None else now_ns())
        payload.setdefault("run_id", self.run_id)
        payload.setdefault("experiment", self.experiment)
        payload.setdefault("condition", self.condition)
        payload.setdefault("source", self.source)
        self._file.write(json.dumps(payload, sort_keys=True, separators=(",", ":")) + "\n")

    def flush(self) -> None:
        self._file.flush()

    def close(self) -> None:
        self._file.flush()
        self._file.close()

    def __enter__(self) -> "NdjsonEventWriter":
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()

File: rust/src/experiments/metrics.rs
//! Metric writers and a monotonic-anchored clock for HORUS benchmark runs.
//!
//! Mirrors `horus.experiments.metrics`: a wall-anchored monotonic clock (so
//! timestamps share an epoch with Quest/Unity wall time but never step
//! backwards on a VM), a newline-delimited JSON event writer, and a CSV metric
//! writer with stable common fields.

use once_cell::sync::Lazy;
use serde_json::{Map, Value};
use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::{Path, PathBuf};
use std::time::{Instant, SystemTime, UNIX_EPOCH};

pub const COMMON_FIELDS: [&str; 4] = ["timestamp_ns", "run_id", "experiment", "condition"];

struct ClockAnchor {
    wall_ns: u128,
    mono: Instant,
}

static CLOCK: Lazy<ClockAnchor> = Lazy::new(|| ClockAnchor {
    wall_ns: SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos(),
    mono: Instant::now(),
});

/// Unix-epoch-like timestamp derived from a monotonic clock, anchored once to
/// wall time. Stays monotonic within the process even if wall time steps.
pub fn now_ns() -> u128 {
    CLOCK.wall_ns + CLOCK.mono.elapsed().as_nanos()
}

/// Write event records as newline-delimited JSON.
pub struct NdjsonEventWriter {
    writer: BufWriter<File>,
    run_id: String,
    experiment: String,
    condition: String,
    source: String,
}

impl NdjsonEventWriter {
    pub fn create(
        path: impl AsRef<Path>,
        run_id: impl Into<String>,
        experiment: impl Into<String>,
        condition: impl Into<String>,
        source: impl Into<String>,
    ) -> std::io::Result<Self> {
        let path = path.as_ref();
        if let Some(parent) = path.parent() {
            if !parent.as_os_str().is_empty() {
                std::fs::create_dir_all(parent)?;
            }
        }
        Ok(Self {
            writer: BufWriter::new(File::create(path)?),
            run_id: run_id.into(),
            experiment: experiment.into(),
            condition: condition.into(),
            source: source.into(),
        })
    }

    pub fn write(&mut self, event: Map<String, Value>, timestamp_ns: Option<u128>) -> std::io::Result<()> {
        let mut payload = event;
        payload
            .entry("timestamp_ns")
            .or_insert_with(|| Value::from(timestamp_ns.unwrap_or_else(now_ns) as u64));
        payload
            .entry("run_id")
            .or_insert_with(|| Value::from(self.run_id.clone()));
        payload
            .entry("experiment")
            .or_insert_with(|| Value::from(self.experiment.clone()));
        payload
            .entry("condition")
            .or_insert_with(|| Value::from(self.condition.clone()));
        payload
            .entry("source")
            .or_insert_with(|| Value::from(self.source.clone()));
        // Stable key order keeps event logs diff-friendly across languages.
        let ordered: std::collections::BTreeMap<String, Value> = payload.into_iter().collect();
        let line = serde_json::to_string(&ordered)?;
        self.writer.write_all(line.as_bytes())?;
        self.writer.write_all(b"\n")
    }

    pub fn flush(&mut self) -> std::io::Result<()> {
        self.writer.flush()
    }
}

/// Append metric rows to a CSV with stable common fields prepended.
pub struct CsvMetricWriter {
    writer: BufWriter<File>,
    fields: Vec<String>,
    run_id: String,
    experiment: String,
    condition: String,
}

impl CsvMetricWriter {
    pub fn create(
        path: impl AsRef<Path>,
        run_id: impl Into<String>,
        experiment: impl Into<String>,
        condition: impl Into<String>,
        fieldnames: &[&str],
    ) -> std::io::Result<Self> {
        let path: PathBuf = path.as_ref().to_path_buf();
        if let Some(parent) = path.parent() {
            if !parent.as_os_str().is_empty() {
                std::fs::create_dir_all(parent)?;
            }
        }
        let mut fields: Vec<String> = COMMON_FIELDS.iter().map(|s| s.to_string()).collect();
        for field in fieldnames {
            if !fields.iter().any(|f| f == field) {
                fields.push((*field).to_string());
            }
        }
        let mut writer = BufWriter::new(File::create(&path)?);
        writeln!(writer, "{}", fields.join(","))?;
        Ok(Self {
            writer,
            fields,
            run_id: run_id.into(),
            experiment: experiment.into(),
            condition: condition.into(),
        })
    }

    pub fn write(&mut self, row: &Map<String, Value>, timestamp_ns: Option<u128>) -> std::io::Result<()> {
        let cells: Vec<String> = self
            .fields
            .iter()
            .map(|field| match field.as_str() {
                "timestamp_ns" => row
                    .get("timestamp_ns")
                    .map(csv_cell)
                    .unwrap_or_else(|| (timestamp_ns.unwrap_or_else(now_ns) as u64).to_string()),
                "run_id" => row.get("run_id").map(csv_cell).unwrap_or_else(|| self.run_id.clone()),
                "experiment" => row
                    .get("experiment")
                    .map(csv_cell)
                    .unwrap_or_else(|| self.experiment.clone()),
                "condition" => row
                    .get("condition")
                    .map(csv_cell)
                    .unwrap_or_else(|| self.condition.clone()),
                other => row.get(other).map(csv_cell).unwrap_or_default(),
            })
            .collect();
        writeln!(self.writer, "{}", cells.join(","))
    }

    pub fn flush(&mut self) -> std::io::Result<()> {
        self.writer.flush()
    }
}

fn csv_cell(value: &Value) -> String {
    let raw = match value {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    };
    if raw.contains(',') || raw.contains('"') || raw.contains('\n') {
        format!("\"{}\"", raw.replace('"', "\"\""))
    } else {
        raw
    }
}
